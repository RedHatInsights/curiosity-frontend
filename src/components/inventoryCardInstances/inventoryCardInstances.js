import React from 'react';
import {
  useGetInstancesInventory,
  useInventoryCardActionsInstances,
  useOnPageInstances,
  useOnColumnSortInstances,
  useParseInstancesFiltersSettings
} from './inventoryCardInstancesContext';
import { InventoryCard } from '../inventoryCard/inventoryCard';
import { helpers } from '../../common';

/**
 * @memberof Components
 * @module InventoryCardInstances
 * @property {module} InventoryCardInstancesContext
 */

/**
 * An instances' system inventory component.
 *
 * @param {object} props
 * @param {boolean} [props.isDisabled=helpers.UI_DISABLED_TABLE_INSTANCES]
 * @param {useGetInstancesInventory} [props.useGetInventory=useGetInstancesInventory]
 * @param {useInventoryCardActionsInstances} [props.useInventoryCardActions=useInventoryCardActionsInstances]
 * @param {useOnPageInstances} [props.useOnPage=useOnPageInstances]
 * @param {useOnColumnSortInstances} [props.useOnColumnSort=useOnColumnSortInstances]
 * @param {useParseInstancesFiltersSettings} [props.useParseFiltersSettings=useParseInstancesFiltersSettings]
 * @returns {JSX.Element}
 */
const InventoryCardInstances = ({
  isDisabled = helpers.UI_DISABLED_TABLE_INSTANCES,
  useGetInventory = useGetInstancesInventory,
  useInventoryCardActions = useInventoryCardActionsInstances,
  useOnPage = useOnPageInstances,
  useOnColumnSort = useOnColumnSortInstances,
  useParseFiltersSettings = useParseInstancesFiltersSettings
}) => (
  <InventoryCard
    isDisabled={isDisabled}
    useGetInventory={useGetInventory}
    useInventoryCardActions={useInventoryCardActions}
    useOnPage={useOnPage}
    useOnColumnSort={useOnColumnSort}
    useParseFiltersSettings={useParseFiltersSettings}
  />
);

export { InventoryCardInstances as default, InventoryCardInstances };
