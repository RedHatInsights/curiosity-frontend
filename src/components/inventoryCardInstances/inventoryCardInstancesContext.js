import React, { useMemo } from 'react';
import { useShallowCompareEffect } from 'react-use';
import { ToolbarItem } from '@patternfly/react-core';
import { reduxActions, reduxTypes, storeHooks } from '../../redux';
import { useSession } from '../authentication/authenticationContext';
import {
  useProduct,
  useProductInventoryGuestsConfig,
  useProductInventoryHostsConfig,
  useProductInventoryHostsQuery
} from '../productView/productViewContext';
import {
  RHSM_API_QUERY_INVENTORY_SORT_DIRECTION_TYPES as SORT_DIRECTION_TYPES,
  RHSM_API_QUERY_INVENTORY_SORT_TYPES as SORT_TYPES,
  RHSM_API_QUERY_SET_TYPES
} from '../../services/rhsm/rhsmConstants';
import { helpers } from '../../common';
import { inventoryCardHelpers } from '../inventoryCard/inventoryCardHelpers';
import { tableHelpers } from '../table/table';
import { toolbarFieldOptions } from '../toolbar/toolbarFieldSelectCategory';
import { InventoryGuests } from '../inventoryGuests/inventoryGuests'; // eslint-disable-line

/**
 * @memberof InventoryCardInstances
 * @module InventoryCardInstancesContext
 */

/**
 * Parse filters settings for context.
 *
 * @param {object} options
 * @param {boolean} options.isDisabled
 * @param {Function} options.useProduct
 * @param {Function} options.useProductConfig
 * @returns {{settings: {}, columnCountAndWidths: {count: number, widths: Array}, filters: Array}}
 */
const useParseInstancesFiltersSettings = ({
  isDisabled = false,
  useProduct: useAliasProduct = useProduct,
  useProductConfig: useAliasProductConfig = useProductInventoryHostsConfig,
  useProductGuestsConfig: useAliasProductGuestsConfig = useProductInventoryGuestsConfig
} = {}) => {
  const { productId } = useAliasProduct();
  const { filters = [], settings = {} } = useAliasProductConfig();
  const { filters: guestFilters = [] } = useAliasProductGuestsConfig();

  return useMemo(() => {
    if (isDisabled) {
      return undefined;
    }
    return inventoryCardHelpers.normalizeInventorySettings({
      filters,
      guestFilters,
      settings,
      productId
    });
  }, [filters, guestFilters, isDisabled, settings, productId]);
};

/**
 * Parse selector response for consuming components.
 *
 * @param {object} options
 * @param {string} options.storeRef
 * @param {Function} options.useParseFiltersSettings
 * @param {Function} options.useProduct
 * @param {Function} options.useProductInventoryQuery
 * @param {Function} options.useSelectorsResponse
 * @param {Function} options.useSession
 * @returns {{pending: boolean, fulfilled: boolean, error: boolean, resultsColumnCountAndWidths: {count: number,
 *     widths: Array}, dataSetColumnHeaders: Array, resultsPerPage: number, resultsOffset: number, dataSetRows: Array,
 *     resultsCount: number}}
 */
const useSelectorInstances = ({
  storeRef = 'instancesInventory',
  useParseFiltersSettings: useAliasParseFiltersSettings = useParseInstancesFiltersSettings,
  useProduct: useAliasProduct = useProduct,
  useProductInventoryQuery: useAliasProductInventoryQuery = useProductInventoryHostsQuery,
  useSelectorsResponse: useAliasSelectorsResponse = storeHooks.reactRedux.useSelectorsResponse,
  useSession: useAliasSession = useSession
} = {}) => {
  const { productId } = useAliasProduct();
  const session = useAliasSession();
  const query = useAliasProductInventoryQuery();
  const { columnCountAndWidths, filters, isGuestFiltersDisabled, settings } = useAliasParseFiltersSettings();
  const response = useAliasSelectorsResponse(({ inventory }) => inventory?.[storeRef]?.[productId]);

  const { pending, cancelled, data, ...restResponse } = response;
  const updatedPending = pending || cancelled || false;
  let parsedData;

  if (response?.fulfilled) {
    const updatedData = (data?.length === 1 && data[0]) || data || {};
    parsedData = inventoryCardHelpers.parseInventoryResponse({
      data: updatedData,
      filters,
      GuestComponent: InventoryGuests,
      isGuestFiltersDisabled,
      query,
      session,
      settings
    });
  }

  return {
    ...restResponse,
    pending: updatedPending,
    resultsColumnCountAndWidths: columnCountAndWidths,
    ...parsedData
  };
};

/**
 * Combine service call, Redux, and inventory selector response.
 *
 * @param {object} options
 * @param {boolean} options.isDisabled
 * @param {Function} options.getInventory
 * @param {Function} options.useDispatch
 * @param {Function} options.useProduct
 * @param {Function} options.useProductInventoryQuery
 * @param {Function} options.useSelector
 * @returns {{pending: boolean, fulfilled: boolean, error: boolean, resultsColumnCountAndWidths: {count: number,
 *     widths: Array}, dataSetColumnHeaders: Array, resultsPerPage: number, resultsOffset: number, dataSetRows: Array,
 *     resultsCount: number}}
 */
const useGetInstancesInventory = ({
  isDisabled = false,
  getInventory = reduxActions.rhsm.getInstancesInventory,
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useProduct: useAliasProduct = useProduct,
  useProductInventoryQuery: useAliasProductInventoryQuery = useProductInventoryHostsQuery,
  useSelector: useAliasSelector = useSelectorInstances
} = {}) => {
  const { productId } = useAliasProduct();
  const query = useAliasProductInventoryQuery();
  const dispatch = useAliasDispatch();
  const response = useAliasSelector();

  useShallowCompareEffect(() => {
    if (!isDisabled) {
      getInventory(productId, query)(dispatch);
    }
  }, [isDisabled, productId, query]);

  return response;
};

/**
 * Return a component list for a configurable inventoryCard action toolbar.
 * Allow the "content" prop to receive inventory data for display via callback.
 *
 * @param {object} options
 * @param {Array} options.categoryOptions
 * @param {Function} options.useSelector
 * @param {Function} options.useProductConfig
 * @returns {Array}
 */
const useInventoryCardActionsInstances = ({
  categoryOptions = toolbarFieldOptions,
  useSelector: useAliasSelector = useSelectorInstances,
  useProductConfig: useAliasProductConfig = useProductInventoryHostsConfig
} = {}) => {
  const results = useAliasSelector();
  const { pending, resultsCount } = results;
  const { settings = {} } = useAliasProductConfig();
  const { actions } = settings;

  return useMemo(
    () =>
      actions?.map(({ id, content, ...actionProps }) => {
        const option = categoryOptions.find(({ value: categoryOptionValue }) => id === categoryOptionValue);
        const { component: OptionComponent } = option || {};

        return (
          (OptionComponent && (
            <ToolbarItem key={`option-${id}`}>
              <OptionComponent isFilter={false} {...actionProps} />
            </ToolbarItem>
          )) ||
          (content && !pending && resultsCount && (
            <ToolbarItem key={id || helpers.generateId()}>
              {typeof content === 'function' ? content({ data: results }) : content}
            </ToolbarItem>
          )) ||
          null
        );
      }),
    [actions, categoryOptions, results, resultsCount, pending]
  );
};

/**
 * An onPage callback for inventory.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @param {Function} options.useProduct
 * @returns {Function}
 */
const useOnPageInstances = ({
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useProduct: useAliasProduct = useProduct
} = {}) => {
  const { productId } = useAliasProduct();
  const dispatch = useAliasDispatch();

  /**
   * On event update state for inventory.
   *
   * @event onPage
   * @param {object} params
   * @param {number} params.offset
   * @param {number} params.perPage
   * @returns {void}
   */
  return ({ offset, perPage }) => {
    dispatch([
      {
        type: reduxTypes.query.SET_QUERY_RHSM_HOSTS_INVENTORY_TYPES[RHSM_API_QUERY_SET_TYPES.OFFSET],
        viewId: productId,
        [RHSM_API_QUERY_SET_TYPES.OFFSET]: offset
      },
      {
        type: reduxTypes.query.SET_QUERY_RHSM_HOSTS_INVENTORY_TYPES[RHSM_API_QUERY_SET_TYPES.LIMIT],
        viewId: productId,
        [RHSM_API_QUERY_SET_TYPES.LIMIT]: perPage
      }
    ]);
  };
};

/**
 * An onColumnSort callback for inventory.
 *
 * @param {object} options
 * @param {object} options.sortColumns
 * @param {Function} options.useDispatch
 * @param {Function} options.useProduct
 * @returns {Function}
 */
const useOnColumnSortInstances = ({
  sortColumns = SORT_TYPES,
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useProduct: useAliasProduct = useProduct
} = {}) => {
  const { productId } = useAliasProduct();
  const dispatch = useAliasDispatch();

  /**
   * On event update state for inventory.
   *
   * @event onColumnSort
   * @param {object} params
   * @param {string} params.direction
   * @param {object} params.data
   * @returns {void}
   */
  return ({ direction, data = {} }) => {
    const { metric: id } = data;
    const updatedSortColumn = Object.values(sortColumns).find(value => value === id);
    let updatedDirection;

    if (!updatedSortColumn) {
      if (helpers.DEV_MODE || helpers.REVIEW_MODE) {
        console.warn(`Sorting can only be performed on select fields, confirm field ${id} is allowed.`);
      }
      return;
    }

    switch (direction) {
      case tableHelpers.SortByDirectionVariant.desc:
        updatedDirection = SORT_DIRECTION_TYPES.DESCENDING;
        break;
      default:
        updatedDirection = SORT_DIRECTION_TYPES.ASCENDING;
        break;
    }

    dispatch([
      {
        type: reduxTypes.query.SET_QUERY_RHSM_HOSTS_INVENTORY_TYPES[RHSM_API_QUERY_SET_TYPES.DIRECTION],
        viewId: productId,
        [RHSM_API_QUERY_SET_TYPES.DIRECTION]: updatedDirection
      },
      {
        type: reduxTypes.query.SET_QUERY_RHSM_HOSTS_INVENTORY_TYPES[RHSM_API_QUERY_SET_TYPES.SORT],
        viewId: productId,
        [RHSM_API_QUERY_SET_TYPES.SORT]: updatedSortColumn
      }
    ]);
  };
};

const context = {
  useGetInstancesInventory,
  useInventoryCardActionsInstances,
  useOnPageInstances,
  useOnColumnSortInstances,
  useParseInstancesFiltersSettings,
  useSelectorInstances
};

export {
  context as default,
  context,
  useGetInstancesInventory,
  useInventoryCardActionsInstances,
  useOnPageInstances,
  useOnColumnSortInstances,
  useParseInstancesFiltersSettings,
  useSelectorInstances
};
