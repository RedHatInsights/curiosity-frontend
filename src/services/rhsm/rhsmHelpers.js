import { RHSM_API_QUERY_SET_TYPES } from './rhsmConstants';

/**
 * ToDo: remove filterArchitectureVariant when the API supports architecture, variant params
 */
/**
 * Patch for returning a made up API architecture, variant param as a product ID
 *
 * @param {string} id
 * @param {object} params
 * @returns {string}
 */
const filterArchitectureVariant = (id, params = {}) => {
  const updatedId = id;

  if (params?.[RHSM_API_QUERY_SET_TYPES.ARCHITECTURE]?.length) {
    return params?.[RHSM_API_QUERY_SET_TYPES.ARCHITECTURE];
  }

  if (params?.[RHSM_API_QUERY_SET_TYPES.VARIANT]?.length) {
    return params?.[RHSM_API_QUERY_SET_TYPES.VARIANT];
  }

  return updatedId;
};

const rhsmHelpers = {
  filterArchitectureVariant
};

export { rhsmHelpers as default, rhsmHelpers, filterArchitectureVariant };
