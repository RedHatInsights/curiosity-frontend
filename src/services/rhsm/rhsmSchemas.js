import JoiBase from 'joi';
import JoiDate from '@joi/date';
import { schemaResponse } from '../common/helpers';
import { rhsmConstants } from './rhsmConstants';

/**
 * @memberof Rhsm
 * @module RhsmSchemas
 */

/**
 * Extend Joi with date.
 *
 * @type {*}
 */
const Joi = JoiBase.extend(JoiDate);

/**
 * Error response item.
 *
 * @type {*} Joi schema
 */
const errorItem = Joi.object({
  code: Joi.string().default(null),
  detail: Joi.string().default(null)
}).unknown(true);

/**
 * Error response.
 *
 * @type {*} Joi schema
 */
const errorResponseSchema = Joi.object()
  .keys({
    errors: Joi.array().items(errorItem).default([])
  })
  .unknown(true);

const linksSchema = Joi.object();

/**
 * RHSM base response meta field.
 *
 * @type {*} Joi schema
 */
const metaResponseSchema = Joi.object()
  .keys({
    count: Joi.number().integer().default(0),
    product: Joi.string().valid(...Object.values(rhsmConstants.RHSM_API_PATH_PRODUCT_TYPES))
  })
  .unknown(true);

/**
 * Capacity response meta field.
 *
 * @type {*} Joi schema
 */
const capacityMetaSchema = metaResponseSchema
  .keys({
    metric_id: Joi.string().valid(...Object.values(rhsmConstants.RHSM_API_PATH_METRIC_TYPES))
  })
  .unknown(true);

/**
 * Capacity response item.
 *
 * @type {*} Joi schema
 */
const capacityItem = Joi.object({
  date: Joi.date().utc().allow(null),
  has_data: Joi.boolean().optional().allow(null),
  has_infinite_quantity: Joi.boolean().optional().allow(null),
  value: Joi.number().allow(null).default(0)
})
  .unknown(true)
  .default();

/**
 * Capacity response.
 *
 * @type {*} Joi schema
 */
const capacityResponseSchema = Joi.object().keys({
  data: Joi.array().items(capacityItem).default([]),
  links: linksSchema.default({}),
  meta: capacityMetaSchema.default({})
});

/**
 * Guests response meta field.
 *
 * @type {*} Joi schema
 */
const guestsMetaSchema = Joi.object()
  .keys({
    count: Joi.number().integer().default(0)
  })
  .unknown(true);

/**
 * Guests response item.
 *
 * @type {*} Joi schema
 */
const guestsItem = Joi.object({
  inventory_id: Joi.string().optional().allow(null),
  display_name: Joi.string().optional().allow(null),
  subscription_manager_id: Joi.string().optional().allow(null),
  last_seen: Joi.date().utc().allow(null)
})
  .unknown(true)
  .default();

/**
 * Guests response.
 *
 * @type {*} Joi schema
 */
const guestsResponseSchema = Joi.object().keys({
  data: Joi.array().items(guestsItem).default([]),
  links: linksSchema.default({}),
  meta: guestsMetaSchema.default({})
});

/**
 * Instances response meta field.
 *
 * @type {*} Joi schema
 */
const instancesMetaSchema = metaResponseSchema
  .keys({
    count: Joi.number().integer().default(0),
    measurements: Joi.array()
      .items(Joi.string().valid(...Object.values(rhsmConstants.RHSM_API_PATH_METRIC_TYPES)))
      .default([])
  })
  .unknown(true);

/**
 * Instances response item.
 *
 * @type {*} Joi schema
 */
const instancesItem = Joi.object({
  instance_id: Joi.string().optional().allow(null),
  category: Joi.string().lowercase().optional().allow(null),
  cloud_provider: Joi.string().lowercase().optional().allow(null, ''),
  display_name: Joi.string().optional().allow(null),
  billing_provider: Joi.string().lowercase().optional().allow(null, ''),
  billing_account_id: Joi.string().optional().allow(null),
  measurements: Joi.array().default([]),
  number_of_guests: Joi.number().integer().default(0),
  subscription_manager_id: Joi.string().optional().allow(null),
  last_seen: Joi.date().utc().allow(null)
})
  .unknown(true)
  .default();

/**
 * Instances response.
 *
 * @type {*} Joi schema
 */
const instancesResponseSchema = Joi.object().keys({
  data: Joi.array().items(instancesItem).default([]),
  links: linksSchema.default({}),
  meta: instancesMetaSchema.default({})
});

/**
 * Subscriptions response meta field.
 *
 * @type {*} Joi schema
 */
const subscriptionsMetaSchema = metaResponseSchema
  .keys({
    subscription_type: Joi.string().valid(null, ...Object.values(rhsmConstants.RHSM_API_RESPONSE_SUBSCRIPTION_TYPES))
  })
  .unknown(true);

/**
 * Subscriptions response item.
 *
 * @type {*} Joi schema
 */
const subscriptionsItem = Joi.object({
  has_infinite_quantity: Joi.boolean().optional().allow(null),
  next_event_date: Joi.date().utc().allow(null),
  product_name: Joi.string().optional().allow(null),
  quantity: Joi.number().allow(null).default(0),
  service_level: Joi.string().valid(...Object.values(rhsmConstants.RHSM_API_RESPONSE_SLA_TYPES)),
  total_capacity: Joi.number().allow(null).default(0),
  uom: Joi.string()
    .lowercase()
    .valid(...Object.values(rhsmConstants.RHSM_API_RESPONSE_UOM_TYPES))
})
  .unknown(true)
  .default();

/**
 * Subscriptions response.
 *
 * @type {*} Joi schema
 */
const subscriptionsResponseSchema = Joi.object().keys({
  data: Joi.array().items(subscriptionsItem).default([]),
  links: linksSchema.default({}),
  meta: subscriptionsMetaSchema.default({})
});

/**
 * Tally and capacity metric response item.
 *
 * @type {*} Joi schema
 */
const tallyItem = Joi.object({
  date: Joi.date().utc().allow(null),
  has_data: Joi.boolean().optional().allow(null),
  value: Joi.number().allow(null).default(0)
})
  .unknown(true)
  .default();

/**
 * Tally response meta field.
 *
 * @type {*} Joi schema
 */
const tallyMetaSchema = metaResponseSchema
  .keys({
    has_cloudigrade_data: Joi.boolean().optional().allow(null),
    has_cloudigrade_mismatch: Joi.boolean().optional().allow(null),
    metric_id: Joi.string().valid(...Object.values(rhsmConstants.RHSM_API_PATH_METRIC_TYPES)),
    total_monthly: tallyItem
  })
  .unknown(true);

/**
 * Tally response.
 *
 * @type {*} Joi schema
 */
const tallyResponseSchema = Joi.object().keys({
  data: Joi.array().items(tallyItem).default([]),
  links: linksSchema.default({}),
  meta: tallyMetaSchema.default({})
});

const rhsmSchemas = {
  capacity: response => schemaResponse({ response, schema: capacityResponseSchema, id: 'RHSM capacity' }),
  errors: response => schemaResponse({ response, schema: errorResponseSchema, id: 'RHSM errors' }),
  guests: response => schemaResponse({ response, schema: guestsResponseSchema, id: 'RHSM guests' }),
  instances: response => schemaResponse({ response, schema: instancesResponseSchema, id: 'RHSM instances' }),
  subscriptions: response =>
    schemaResponse({ response, schema: subscriptionsResponseSchema, id: 'RHSM subscriptions' }),
  tally: response => schemaResponse({ response, schema: tallyResponseSchema, id: 'RHSM tally' })
};

export { rhsmSchemas as default, rhsmSchemas };
