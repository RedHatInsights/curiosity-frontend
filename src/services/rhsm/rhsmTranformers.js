import {
  RHSM_API_RESPONSE_TALLY_DATA_TYPES as TALLY_DATA_TYPES,
  RHSM_API_RESPONSE_TALLY_META_TYPES as TALLY_META_TYPES,
  rhsmConstants
} from './rhsmConstants';

const rhsmTally = response => {
  const updatedResponse = {};
  const { [rhsmConstants.RHSM_API_RESPONSE_DATA]: data = [], [rhsmConstants.RHSM_API_RESPONSE_META]: meta = {} } =
    response || {};

  updatedResponse.data = data.map(
    (
      { [TALLY_DATA_TYPES.DATE]: date, [TALLY_DATA_TYPES.VALUE]: value, [TALLY_DATA_TYPES.HAS_DATA]: hasData },
      index
    ) => ({
      x: index,
      y: value,
      date,
      hasData
    })
  );

  updatedResponse.meta = {
    count: meta[TALLY_META_TYPES.COUNT],
    metricId: meta[TALLY_META_TYPES.METRIC_ID],
    productId: meta[TALLY_META_TYPES.PRODUCT],
    totalMonthlyDate: meta[TALLY_META_TYPES.TOTAL_MONTHLY]?.[TALLY_META_TYPES.DATE],
    totalMonthlyHasData: meta[TALLY_META_TYPES.TOTAL_MONTHLY]?.[TALLY_META_TYPES.HAS_DATA],
    totalMonthlyValue: meta[TALLY_META_TYPES.TOTAL_MONTHLY]?.[TALLY_META_TYPES.VALUE]
  };

  return updatedResponse;
};

const rhsmTranformers = {
  tally: rhsmTally
};

export { rhsmTranformers as default, rhsmTranformers, rhsmTally };
