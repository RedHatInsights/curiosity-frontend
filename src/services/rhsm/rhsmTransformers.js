import moment from 'moment';
import {
  RHSM_API_RESPONSE_INSTANCES_DATA_TYPES as INSTANCES_DATA_TYPES,
  RHSM_API_RESPONSE_INSTANCES_META_TYPES as INSTANCES_META_TYPES,
  RHSM_API_RESPONSE_TALLY_DATA_TYPES as TALLY_DATA_TYPES,
  RHSM_API_RESPONSE_TALLY_META_TYPES as TALLY_META_TYPES,
  rhsmConstants
} from './rhsmConstants';
import { dateHelpers } from '../../common';

/**
 * FixMe: If RHSM Instances is deprecating Hosts we're missing a property, number_of_guests
 */
/**
 * Parse RHSM instances response for caching.
 *
 * @param {object} response
 * @returns {object}
 */
const rhsmInstances = response => {
  const updatedResponse = {};
  const { [rhsmConstants.RHSM_API_RESPONSE_DATA]: data = [], [rhsmConstants.RHSM_API_RESPONSE_META]: meta = {} } =
    response || {};
  const metaMeasurements = meta[INSTANCES_META_TYPES.MEASUREMENTS];

  updatedResponse.data = data.map(
    ({
      [INSTANCES_DATA_TYPES.MEASUREMENTS]: measurements,
      [INSTANCES_DATA_TYPES.SUBSCRIPTION_MANAGER_ID]: subscriptionManagerId,
      [INSTANCES_DATA_TYPES.NUMBER_OF_GUESTS]: numberOfGuests,
      ...dataResponse
    }) => {
      const updatedData = {
        numberOfGuests,
        subscriptionManagerId,
        ...dataResponse
      };

      metaMeasurements?.forEach((measurement, index) => {
        updatedData[measurement] = measurements[index];
      });

      return updatedData;
    }
  );

  updatedResponse.meta = {
    count: meta[INSTANCES_META_TYPES.COUNT],
    productId: meta[INSTANCES_META_TYPES.PRODUCT]
  };

  return updatedResponse;
};

/**
 * Parse RHSM tally response for caching.
 *
 * @param {object} response
 * @returns {object}
 */
const rhsmTally = response => {
  const updatedResponse = {};
  const { [rhsmConstants.RHSM_API_RESPONSE_DATA]: data = [], [rhsmConstants.RHSM_API_RESPONSE_META]: meta = {} } =
    response || {};
  const currentDay = moment.utc(dateHelpers.getCurrentDate()).format('MM-D-YYYY');

  updatedResponse.data = data.map(
    (
      { [TALLY_DATA_TYPES.DATE]: date, [TALLY_DATA_TYPES.VALUE]: value, [TALLY_DATA_TYPES.HAS_DATA]: hasData },
      index
    ) => ({
      x: index,
      y: value,
      date,
      hasData,
      isCurrentDate: moment.utc(date).format('MM-D-YYYY') === currentDay
    })
  );

  updatedResponse.meta = {
    count: meta[TALLY_META_TYPES.COUNT],
    cloudigradeHasMismatch: meta[TALLY_META_TYPES.HAS_CLOUDIGRADE_MISMATCH],
    metricId: meta[TALLY_META_TYPES.METRIC_ID],
    productId: meta[TALLY_META_TYPES.PRODUCT],
    totalMonthlyDate: meta[TALLY_META_TYPES.TOTAL_MONTHLY]?.[TALLY_META_TYPES.DATE],
    totalMonthlyHasData: meta[TALLY_META_TYPES.TOTAL_MONTHLY]?.[TALLY_META_TYPES.HAS_DATA],
    totalMonthlyValue: meta[TALLY_META_TYPES.TOTAL_MONTHLY]?.[TALLY_META_TYPES.VALUE]
  };

  return updatedResponse;
};

const rhsmTransformers = {
  instances: rhsmInstances,
  tally: rhsmTally
};

export { rhsmTransformers as default, rhsmTransformers, rhsmInstances, rhsmTally };
