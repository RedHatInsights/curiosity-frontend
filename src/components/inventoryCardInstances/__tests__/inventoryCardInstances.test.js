import React from 'react';
import { InventoryCardInstances } from '../inventoryCardInstances';
import { RHSM_API_QUERY_SET_TYPES } from '../../../services/rhsm/rhsmConstants';

describe('InventoryCardInstances Component', () => {
  it('should render a basic component', async () => {
    const props = {
      useProductInventoryConfig: () => ({ filters: [], settings: {} }),
      useProductInventoryQuery: () => ({
        [RHSM_API_QUERY_SET_TYPES.LIMIT]: 10,
        [RHSM_API_QUERY_SET_TYPES.OFFSET]: 0
      })
    };

    const component = await shallowComponent(<InventoryCardInstances {...props} />);
    expect(component).toMatchSnapshot('basic');
  });
});
