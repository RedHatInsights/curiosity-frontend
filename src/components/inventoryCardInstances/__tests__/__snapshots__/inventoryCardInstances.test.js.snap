// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InventoryCardInstances Component should render a basic component: basic 1`] = `
<InventoryCard
  isDisabled={false}
  t={[Function]}
  useGetInventory={[Function]}
  useInventoryCardActions={[Function]}
  useOnColumnSort={[Function]}
  useOnPage={[Function]}
  useParseFiltersSettings={[Function]}
  useProductInventoryConfig={[Function]}
  useProductInventoryQuery={[Function]}
/>
`;
